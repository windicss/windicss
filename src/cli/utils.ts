import fs from 'fs';
import path from 'path';
import glob from 'glob';
import minimatch from 'minimatch';

export function globArray(patterns: string[], options?: glob.IOptions): string[] {
  const list: string[] = [];

  patterns.forEach(pattern => {
    if (pattern[0] === '!') {
      let i = list.length-1;
      while(i > -1) {
        if (!minimatch(list[i], pattern)) {
          list.splice(i, 1);
        }
        i--;
      }

    }
    else {
      const newList = glob.sync(pattern, options);
      newList.forEach(item => {
        if (list.indexOf(item)===-1) {
          list.push(item);
        }
      });
    }
  });

  return list;
}

export function getVersion(): string {
  // eslint-disable-next-line quotes
  return `__NAME__ __VERSION__`; // replace by rollup
}

export function fuzzy(content: string): string[] {
  return content.match(/([^<>"'`\s]*[^<>"'`\s:])|([^<>"'`\s.(){}[\]#=%]*[^<>"'`\s.(){}[\]#=%:])/g) ?? [];
}

export function generateTemplate(
  folder: string,
  outputPath = 'windi.css'
): { html: string; css: string } {
  if (!(fs.existsSync(folder) && fs.lstatSync(folder).isDirectory())) {
    fs.mkdirSync(folder);
    if (!fs.existsSync(folder))
      throw new Error(`Folder ${folder} creation failed.`);
  }
  folder = path.resolve(folder);
  const template = `<!DOCTYPE html>
  <html lang="en">

  <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>${path.basename(folder)}</title>
      <link rel="stylesheet" href="${outputPath}">
  </head>

  <body class="bg-gray-100">
      <div class="container mx-auto flex flex-col justify-center items-center h-screen">
          <div class="lg:flex shadow rounded-lg">
              <div class="bg-blue-500 rounded-t-lg lg:rounded-tr-none lg:rounded-l-lg lg:w-4/12 py-4 h-full flex flex-col justify-center">
                  <div class="text-center tracking-wide">
                      <div class="text-white font-bold text-8xl lg:text-4xl">24</div>
                      <div class="text-white font-normal text-4xl pt-4 lg:pt-0 lg:text-2xl">Sept</div>
                  </div>
              </div>
              <div class="w-full px-1 bg-white py-5 lg:px-2 lg:py-2 tracking-wide">
                  <div class="flex flex-row lg:justify-start justify-center">
                      <div class="text-gray-700 font-light text-sm text-center lg:text-left px-2">
                          1:30 PM
                      </div>
                      <div class="text-gray-700 font-light text-sm text-center lg:text-left px-2">
                          Organiser : IHC
                      </div>
                  </div>
                  <div class="text-gray-700 text-xl pb-1 text-center lg:text-left px-2">
                      International Conference Dubai
                  </div>

                  <div class="text-gray-800 font-light text-sm pt-1 text-center lg:text-left px-2">
                      A-142/1, A-142, Ganesh Nagar, Tilak Nagar, New Delhi, 110018
                  </div>
              </div>
              <div class="flex flex-row items-center w-full lg:w-1/3 bg-white lg:justify-end justify-center px-2 py-4 lg:px-0 rounded-lg">
                  <span class="tracking-wider text-gray-600 cursor-pointer bg-gray-100 hover:bg-gray-200 px-4 py-2 text-sm rounded-lg leading-loose mx-2">
                      Going
                  </span>
              </div>
          </div>
      </div>
  </body>
  </html>`;
  const inputPath = path.join(folder, 'index.html');
  outputPath = path.join(folder, outputPath);
  fs.writeFileSync(inputPath, template);
  fs.writeFileSync(outputPath, '');
  return { html: inputPath, css: outputPath };
}
