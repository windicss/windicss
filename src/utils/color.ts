import colorString from 'color-string';
import type { Color } from 'color-string';

export function hsl2rgb(h: number, s: number, l: number): [number, number, number] {
  l /= 100;
  if (h >= 360) h %= 360;

  const c = (1 - Math.abs(2 * l - 1)) * (s / 100);
  const x = c * (1 - Math.abs((h / 60) % 2 - 1));
  const m = l - c/2;
  let r = 0;
  let g = 0;
  let b = 0;

  if (0 <= h && h < 60) {
    r = c + m; g = x + m; b = m;
  } else if (60 <= h && h < 120) {
    r = x + m; g = c + m; b = m;
  } else if (120 <= h && h < 180) {
    r = m; g = c + m; b = x + m;
  } else if (180 <= h && h < 240) {
    r = m; g = x + m; b = c + m;
  } else if (240 <= h && h < 300) {
    r = x + m; g = m; b = c + m;
  } else if (300 <= h && h < 360) {
    r = c + m; g = m; b = x + m;
  }

  return [
    Math.round(r * 255),
    Math.round(g * 255),
    Math.round(b * 255),
  ];
}

export function hwb2rgb(h: number, w: number, b: number): [number, number, number] {
  const rgb = hsl2rgb(h, 100, 50);

  for (let i = 0; i < 3; ++i) {
    let c = rgb[i] / 255;

    c *= 1 - w/100 - b/100;
    c += w/100;

    rgb[i] = Math.round(c * 255);
  }

  return rgb;
}

export function toRGBA(color: string): Color | undefined {
  if (/^hsla?/.test(color)) {
    const color_array = colorString.get.hsl(color);
    if (!color_array) return;
    return [...hsl2rgb(color_array[0], color_array[1], color_array[2]), color_array[3]];
  } else if (/^rgba?/.test(color)) {
    const color_array = colorString.get.rgb(color);
    if (!color_array) return;
    return color_array;
  } else if (color.startsWith('hwb')) {
    const color_array = colorString.get.hwb(color);
    if (!color_array) return;
    return [...hwb2rgb(color_array[0], color_array[1], color_array[2]), color_array[3]];
  }
  return colorString.get(color)?.value;
}

export function toRGB(color: string): number[] | undefined {
  const rgba = toRGBA(color);
  if (!rgba) return;
  rgba.pop();
  return rgba;
}

export function toColor(color_string: string) : { color: string, opacity: string } {
  const rgba = toRGBA(color_string);
  const color = rgba ? rgba.slice(0, 3).join(', ') : color_string;
  const opacity = rgba ? rgba[3].toString() : '1';

  return { color, opacity };
}
