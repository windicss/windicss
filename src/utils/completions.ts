import type { Processor } from '../lib';
import { flatColors } from './tools';
const utilities: { [key:string]: string[]} = {
  // Layout
  columns: [
    'columns-${static}',
    'columns-${float}',
    'columns-${size}',
    'columns-${int}xl',
  ],
  container: [
    'container',
  ],
  objectPosition: [
    'object-${static}',
  ],
  inset: [
    'inset-${static}',
    'inset-${float}',
    'inset-${fraction}',
    'inset-${size}',

    'inset-y-${static}',
    'inset-y-${float}',
    'inset-y-${fraction}',
    'inset-y-${size}',

    'inset-x-${static}',
    'inset-x-${float}',
    'inset-x-${fraction}',
    'inset-x-${size}',

    'top-${static}',
    'top-${float}',
    'top-${fraction}',
    'top-${size}',

    'right-${static}',
    'right-${float}',
    'right-${fraction}',
    'right-${size}',

    'bottom-${static}',
    'bottom-${float}',
    'bottom-${fraction}',
    'bottom-${size}',

    'left-${static}',
    'left-${float}',
    'left-${fraction}',
    'left-${size}',
  ],
  zIndex: [
    'z-${static}',
    'z-${int}',
  ],
  // Flexbox
  flex: [
    'flex-${static}',
  ],
  flexGrow: [
    'flex-grow-${static}',
  ],
  flexShrink: [
    'flex-shrink-${static}',
  ],
  order: [
    'order-${static}',
    'order-${int}',
  ],
  // Grid
  gridTemplateColumns: [
    'grid-cols-${static}',
    'grid-cols-${int}',
  ],
  gridTemplateRows: [
    'grid-rows-${static}',
    'grid-rows-${int}',
  ],
  gridColumn: [
    'col-${static}',
    'col-span-${int}',
  ],
  gridColumnEnd: [
    'col-end-${static}',
    'col-end-${int}',
  ],
  gridColumnStart: [
    'col-start-${static}',
    'col-start-${int}',
  ],
  gridRow: [
    'row-${static}',
    'row-span-${int}',
  ],
  gridRowEnd: [
    'row-end-${static}',
    'row-end-${int}',
  ],
  gridRowStart: [
    'row-start-${static}',
    'row-start-${int}',
  ],
  gap: [
    'gap-${static}',
    'gap-x-${static}',
    'gap-y-${static}',

    'gap-${float}',
    'gap-x-${float}',
    'gap-y-${float}',

    'gap-${size}',
    'gap-x-${size}',
    'gap-y-${size}',
  ],
  // Box Alignment
  // Spacing
  padding: [
    'p-${static}',
    'py-${static}',
    'px-${static}',
    'pt-${static}',
    'pr-${static}',
    'pb-${static}',
    'pl-${static}',

    'p-${float}',
    'py-${float}',
    'px-${float}',
    'pt-${float}',
    'pr-${float}',
    'pb-${float}',
    'pl-${float}',

    'p-${size}',
    'py-${size}',
    'px-${size}',
    'pt-${size}',
    'pr-${size}',
    'pb-${size}',
    'pl-${size}',
  ],
  margin: [
    'm-${static}',
    'my-${static}',
    'mx-${static}',
    'mt-${static}',
    'mr-${static}',
    'mb-${static}',
    'ml-${static}',

    'm-${float}',
    'my-${float}',
    'mx-${float}',
    'mt-${float}',
    'mr-${float}',
    'mb-${float}',
    'ml-${float}',

    'm-${size}',
    'my-${size}',
    'mx-${size}',
    'mt-${size}',
    'mr-${size}',
    'mb-${size}',
    'ml-${size}',
  ],
  space: [
    'space-y-${static}',
    'space-y-reverse',
    'space-x-${static}',
    'space-x-reverse',
    'space-y-${float}',
    'space-x-${float}',
  ],
  width: [
    'w-${static}',
    'w-${float}',
    'w-${fraction}',
    'w-${int}xl',
    'w-${size}',
  ],
  minWidth: [
    'min-w-${static}',
    'min-w-${float}',
    'min-w-${fraction}',
    'min-w-${int}xl',
    'min-w-${size}',
  ],
  maxWidth: [
    'max-w-${static}',
    'max-w-${float}',
    'max-w-${fraction}',
    'max-w-${int}xl',
    'max-w-${size}',
  ],
  height: [
    'h-${static}',
    'h-${float}',
    'h-${fraction}',
    'h-${int}xl',
    'h-${size}',
  ],
  minHeight: [
    'min-h-${static}',
    'min-h-${float}',
    'min-h-${fraction}',
    'min-h-${int}xl',
    'min-h-${size}',
  ],
  maxHeight: [
    'max-h-${static}',
    'max-h-${float}',
    'max-h-${fraction}',
    'max-h-${int}xl',
    'max-h-${size}',
  ],
  // Typography
  fontSize: [
    'text-${static}',
    'text-${int}xl',
  ],
  textOpacity: [
    'text-opacity-${static}',
    'text-opacity-${int<=100}',
  ],
  textColor: [
    'text-${color}',
  ],
  fontFamily: [
    'font-${static}',
  ],
  fontWeight: [
    'font-${static}',
    'font-${int}',
  ],
  letterSpacing: [
    'tracking-${static}',
    'tracking-${size}',
  ],
  lineHeight: [
    'leading-${static}',
    'leading-${int}',
    'leading-${size}',
  ],
  listStyleType: [
    'list-${static}',
  ],
  placeholderColor: [
    'placeholder-${color}',
  ],
  placeholderOpacity: [
    'placeholder-opacity-${static}',
    'placeholder-opacity-${int<=100}',
  ],
  // Backgrounds
  backgroundColor: [
    'bg-${color}',
  ],
  backgroundOpacity: [
    'bg-opacity-${static}',
    'bg-opacity-${int<=100}',
  ],
  backgroundPosition: [
    'bg-${static}',
  ],
  backgroundSize: [
    'bg-${static}',
  ],
  backgroundImage: [
    'bg-${static}',
  ],
  gradientColorStops: [
    'from-${color}',
    'via-${color}',
    'to-${color}',
  ],
  // Borders
  borderRadius: [
    'rounded-${static}',
    'rounded-t-${static}',
    'rounded-l-${static}',
    'rounded-r-${static}',
    'rounded-b-${static}',
    'rounded-tl-${static}',
    'rounded-tr-${static}',
    'rounded-br-${static}',
    'rounded-bl-${static}',

    'rounded-${int}xl',
    'rounded-${size}',
    'rounded-t-${int}xl',
    'rounded-t-${size}',
    'rounded-l-${int}xl',
    'rounded-l-${size}',
    'rounded-r-${int}xl',
    'rounded-r-${size}',
    'rounded-b-${int}xl',
    'rounded-b-${size}',
    'rounded-tl-${int}xl',
    'rounded-tl-${size}',
    'rounded-tr-${int}xl',
    'rounded-tr-${size}',
    'rounded-br-${int}xl',
    'rounded-br-${size}',
    'rounded-bl-${int}xl',
    'rounded-bl-${size}',
  ],
  borderWidth: [
    'border-${static}',
    'border-${int}',
    'border-${size}',
    'border-t-${int}',
    'border-t-${size}',
    'border-r-${int}',
    'border-r-${size}',
    'border-b-${int}',
    'border-b-${size}',
    'border-l-${int}',
    'border-l-${size}',
  ],
  borderColor: [
    'border-${color}',
  ],
  borderOpacity: [
    'border-opacity-${static}',
    'border-opacity-${int<=100}',
  ],
  divideWidth: [
    'divide-y-reverse',
    'divide-x-reverse',
    'divide-y-${int}',
    'divide-x-${int}',
  ],
  divideColor: [
    'divide-${color}',
  ],
  divideOpacity: [
    'divide-${static}',
    'divide-opacity-${int<=100}',
  ],
  ringOffsetWidth: [
    'ring-offset-${static}',
    'ring-offset-${int}',
  ],
  ringOffsetColor: [
    'ring-offset-${color}',
  ],
  ringWidth: [
    'ring-${static}',
    'ring-${int}',
  ],
  ringColor: [
    'ring-${color}',
  ],
  ringOpacity: [
    'ring-${static}',
    'ring-opacity-${int<=100}',
  ],
  // Effects
  boxShadow: [
    'shadow-${static}',
  ],
  opacity: [
    'opacity-${static}',
    'opacity-${int<=100}',
  ],
  transition: [
    'transition-${static}',
  ],
  transitionDuration: [
    'duration-${static}',
    'duration-${int}',
  ],
  transitionTimingFunction: [
    'ease-${static}',
  ],
  transitionDelay: [
    'delay-${static}',
    'delay-${int}',
  ],
  animation: [
    'animate-${static}',
  ],
  // Transforms
  transformOrigin: [
    'origin-${static}',
  ],
  scale: [
    'scale-${static}',
    'scale-${int}',
    'scale-x-${static}',
    'scale-x-${int}',
    'scale-y-${static}',
    'scale-y-${int}',
  ],
  rotate: [
    'rotate-${static}',
    'rotate-${float}',
  ],
  translate: [
    'translate-${static}',
    'translate-x-${static}',
    'translate-y-${static}',
    'translate-x-${float}',
    'translate-x-${fraction}',
    'translate-x-${size}',
    'translate-y-${float}',
    'translate-y-${fraction}',
    'translate-y-${size}',
  ],
  skew: [
    'skew-x-${static}',
    'skew-x-${float}',
    'skew-y-${static}',
    'skew-y-${float}',
  ],
  cursor: [
    'cursor-${static}',
  ],
  // Interactivity
  outline: [
    'outline-${static}',
  ],
  outlineColor: [
    'outline-${color}',
    'outline-solid-${color}',
    'outline-dotted-${color}',
  ],
  // SVG
  fill: [
    'fill-${color}',
  ],
  // Stroke
  stroke: [
    'stroke-${color}',
  ],
  strokeWidth: [
    'stroke-${int}',
  ],

  // Plugins
  typography: [
    'prose-sm',
    'prose',
    'prose-lg',
    'prose-xl',
    'prose-2xl',
    'prose-red',
    'prose-yellow',
    'prose-green',
    'prose-blue',
    'prose-indigo',
    'prose-purple',
    'prose-pink',
  ],
  aspectRatio: [
    'aspect-none',
    'aspect-auto',
    'aspect-square',
    'aspect-video',
    'aspect-w-${float}',
    'aspect-h-${float}',
    'aspect-${fraction}',
  ],
  lineClamp: [
    'line-clamp-none',
    'line-clamp-${int}',
  ],
  filter: [
    'filter-${static}',
  ],
  backdropFilter: [
    'backdrop-${static}',
  ],
  blur: [
    'blur-${static}',
    'blur-${float}',
    'blur-${size}',
  ],
};

const negative: { [key:string]: true} = {
  inset: true,
  zIndex: true,
  order: true,
  margin: true,
  space: true,
  letterSpacing: true,
  rotate: true,
  translate: true,
  skew: true,
};

export function generateCompletions(processor: Processor): { static: string[], color: string[], dynamic: string[]} {
  const completions : {
    static: string[],
    color: string[],
    dynamic: string[]
  } = { static: [], color: [], dynamic: [] };
  const colors = flatColors(processor.theme('colors') as {[key:string]:string|{[key:string]:string}});
  for (const [config, list] of Object.entries(utilities)) {
    list.forEach(utility => {
      const mark = utility.search(/\$/);
      if (mark === -1) {
        completions.static.push(utility);
      } else {
        const prefix = utility.slice(0, mark-1);
        const suffix = utility.slice(mark);
        switch(suffix) {
        case '${static}':
          completions.static = completions.static.concat(Object.keys(processor.theme(config, {}) as any).map(i => i === 'DEFAULT' ? prefix : i.charAt(0) === '-' ? `-${prefix}${i}` : `${prefix}-${i}`));
          break;
        case '${color}':
          for (const key of Object.keys(flatColors(processor.theme(config, colors) as any))) {
            if (key !== 'DEFAULT')
              completions.color.push(`${prefix}-${key}`);
          }
          break;
        default:
          completions.dynamic.push(utility);
          if (config in negative) completions.dynamic.push(`-${utility}`);
          break;
        }
      }
    });
  }
  return completions;
}
