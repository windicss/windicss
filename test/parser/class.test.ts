import { ClassParser } from '../../src/utils/parser';

describe('ClassParser', () => {
  it('should remove duplicated classes', () => {
    expect(new ClassParser('font-bold font-bold bg-red-300').parse().length).toEqual(2);
    expect(new ClassParser('font-bold font-bold bg-red-300').parse(false).length).toEqual(3);
  });

  it('parse', () => {
    const classes = 'font-bold text-green-300 dark:font-medium dark:p-4 -sm:border +sm:float-right md: hover:  (bg-black-300 text-gray-200   text-lg dark:(bg-black-300 text-gray-200)) md:text-red-500 hover:  text-red-300 text-green-300 sm:hover:bg-red-500 dark:hover:bg-black-300 sm:dark:hover:bg-gray-300 abc bg-cool-gray-300 bg-hex-fff';
    const parser = new ClassParser(undefined, ':', ['dark', '-sm', '+sm', 'md', 'hover', 'sm']);
    expect(parser.parse().length).toBe(0);
    parser.classNames = classes;
    expect(parser.parse()).toMatchSnapshot('class parser');
  });

  it('parse important', () => {
    const classes = '!text-green-300 font-bold !hover:(p-4 bg-red-500) focus:(!border float-right)';
    const parser = new ClassParser(classes, ':', ['hover', 'focus']);
    expect(parser.parse()).toMatchSnapshot('parse important');
  });

  it('parse alias', () => {
    const classes = 'font-bold *hstack sm:(*vstack text-green-200)';
    const parser = new ClassParser(classes, ':', ['sm']);
    expect(parser.parse()).toMatchSnapshot('parse alias');
  });

  it('parse bad class', () => {
    const classes = 'css`height:calc-(100 float-left w-250px h-full bg-hex-f7f7fa overflow-y-auto height:calc-(100';
    const parser = new ClassParser(classes, ':', []);
    expect(parser.parse()).toMatchSnapshot('bad classes');
  });

  it('parse bad class with quotes', () => {
    const classes = '${ active ? \'text-green-400\' : \'text-orange-400\' }';
    const parser = new ClassParser(classes, ':', []);
    expect(parser.parse()).toMatchSnapshot('bad classes');
  });

  it('parse bad class with bracket', () => {
    const classes = 'xyz() p-5 bg-red-500 xyz-() xy-( xy) bg-yellow-600';
    const parser = new ClassParser(classes, ':', []);
    expect(parser.parse()).toMatchSnapshot('bad bracket');
  });

  it('parse bad class with half bracket', () => {
    const classes = 'abc( text-lg font-bold';
    const parser = new ClassParser(classes, ':', []);
    expect(parser.parse()).toMatchSnapshot('bad half bracket');
  });

  it('parse function with parameter deconstruction', () => {
    const classes = 'function ComponentA({ name } Props) { return div className= grid }';
    const result = new ClassParser(classes, ':', []).parse();

    const findGrid = (els: any[]) => els.find(item => item.raw === 'grid');
    expect(findGrid(result)).toBeTruthy();
    expect(result).toMatchSnapshot('jsx function component');
  });
});
