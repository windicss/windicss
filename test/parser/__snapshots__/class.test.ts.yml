Tools / parse / class parser / 0: |-
  [
    {
      "raw": "font-bold",
      "start": 0,
      "end": 9,
      "variants": [],
      "content": "font-bold",
      "type": "utility",
      "important": false
    },
    {
      "raw": "text-green-300",
      "start": 10,
      "end": 24,
      "variants": [],
      "content": "text-green-300",
      "type": "utility",
      "important": false
    },
    {
      "raw": "dark:font-medium",
      "start": 25,
      "end": 41,
      "variants": [
        "dark"
      ],
      "content": "font-medium",
      "type": "utility",
      "important": false
    },
    {
      "raw": "dark:p-4",
      "start": 42,
      "end": 50,
      "variants": [
        "dark"
      ],
      "content": "p-4",
      "type": "utility",
      "important": false
    },
    {
      "raw": "-sm:border",
      "start": 51,
      "end": 61,
      "variants": [
        "-sm"
      ],
      "content": "border",
      "type": "utility",
      "important": false
    },
    {
      "raw": "+sm:float-right",
      "start": 62,
      "end": 77,
      "variants": [
        "+sm"
      ],
      "content": "float-right",
      "type": "utility",
      "important": false
    },
    {
      "raw": "md: hover:  (bg-black-300 text-gray-200   text-lg dark:(bg-black-300 text-gray-200))",
      "start": 78,
      "end": 162,
      "variants": [
        "md",
        "hover"
      ],
      "content": [
        {
          "raw": "bg-black-300",
          "start": 91,
          "end": 103,
          "variants": [],
          "content": "bg-black-300",
          "type": "utility",
          "important": false
        },
        {
          "raw": "text-gray-200",
          "start": 104,
          "end": 117,
          "variants": [],
          "content": "text-gray-200",
          "type": "utility",
          "important": false
        },
        {
          "raw": "text-lg",
          "start": 120,
          "end": 127,
          "variants": [],
          "content": "text-lg",
          "type": "utility",
          "important": false
        },
        {
          "raw": "dark:(bg-black-300 text-gray-200)",
          "start": 128,
          "end": 161,
          "variants": [
            "dark"
          ],
          "content": [
            {
              "raw": "bg-black-300",
              "start": 134,
              "end": 146,
              "variants": [],
              "content": "bg-black-300",
              "type": "utility",
              "important": false
            },
            {
              "raw": "text-gray-200",
              "start": 147,
              "end": 160,
              "variants": [],
              "content": "text-gray-200",
              "type": "utility",
              "important": false
            }
          ],
          "type": "group",
          "important": false
        }
      ],
      "type": "group",
      "important": false
    },
    {
      "raw": "md:text-red-500",
      "start": 163,
      "end": 178,
      "variants": [
        "md"
      ],
      "content": "text-red-500",
      "type": "utility",
      "important": false
    },
    {
      "raw": "hover:  text-red-300",
      "start": 179,
      "end": 199,
      "variants": [
        "hover"
      ],
      "content": "text-red-300",
      "type": "utility",
      "important": false
    },
    {
      "raw": "sm:hover:bg-red-500",
      "start": 215,
      "end": 234,
      "variants": [
        "sm",
        "hover"
      ],
      "content": "bg-red-500",
      "type": "utility",
      "important": false
    },
    {
      "raw": "dark:hover:bg-black-300",
      "start": 235,
      "end": 258,
      "variants": [
        "dark",
        "hover"
      ],
      "content": "bg-black-300",
      "type": "utility",
      "important": false
    },
    {
      "raw": "sm:dark:hover:bg-gray-300",
      "start": 259,
      "end": 284,
      "variants": [
        "sm",
        "dark",
        "hover"
      ],
      "content": "bg-gray-300",
      "type": "utility",
      "important": false
    },
    {
      "raw": "abc",
      "start": 285,
      "end": 288,
      "variants": [],
      "content": "abc",
      "type": "utility",
      "important": false
    },
    {
      "raw": "bg-cool-gray-300",
      "start": 289,
      "end": 305,
      "variants": [],
      "content": "bg-cool-gray-300",
      "type": "utility",
      "important": false
    },
    {
      "raw": "bg-hex-fff",
      "start": 306,
      "end": 316,
      "variants": [],
      "content": "bg-hex-fff",
      "type": "utility",
      "important": false
    }
  ]
Tools / parse alias / parse alias / 0: |-
  [
    {
      "raw": "font-bold",
      "start": 0,
      "end": 9,
      "variants": [],
      "content": "font-bold",
      "type": "utility",
      "important": false
    },
    {
      "raw": "*hstack",
      "start": 10,
      "end": 17,
      "variants": [],
      "content": "hstack",
      "type": "alias",
      "important": false
    },
    {
      "raw": "sm:(*vstack text-green-200)",
      "start": 18,
      "end": 45,
      "variants": [
        "sm"
      ],
      "content": [
        {
          "raw": "*vstack",
          "start": 22,
          "end": 29,
          "variants": [],
          "content": "vstack",
          "type": "alias",
          "important": false
        },
        {
          "raw": "text-green-200",
          "start": 30,
          "end": 44,
          "variants": [],
          "content": "text-green-200",
          "type": "utility",
          "important": false
        }
      ],
      "type": "group",
      "important": false
    }
  ]
Tools / parse bad class / bad classes / 0: |-
  [
    {
      "raw": "css",
      "start": 0,
      "end": 3,
      "variants": [],
      "content": "css",
      "type": "utility",
      "important": false
    },
    {
      "raw": "height:calc-(100",
      "start": 4,
      "end": 20,
      "variants": [],
      "content": "height:calc-(100",
      "type": "utility",
      "important": false
    },
    {
      "raw": "float-left",
      "start": 21,
      "end": 31,
      "variants": [],
      "content": "float-left",
      "type": "utility",
      "important": false
    },
    {
      "raw": "w-250px",
      "start": 32,
      "end": 39,
      "variants": [],
      "content": "w-250px",
      "type": "utility",
      "important": false
    },
    {
      "raw": "h-full",
      "start": 40,
      "end": 46,
      "variants": [],
      "content": "h-full",
      "type": "utility",
      "important": false
    },
    {
      "raw": "bg-hex-f7f7fa",
      "start": 47,
      "end": 60,
      "variants": [],
      "content": "bg-hex-f7f7fa",
      "type": "utility",
      "important": false
    },
    {
      "raw": "overflow-y-auto",
      "start": 61,
      "end": 76,
      "variants": [],
      "content": "overflow-y-auto",
      "type": "utility",
      "important": false
    },
    {
      "raw": "height:calc-(100)",
      "start": 77,
      "end": 94,
      "variants": [],
      "content": "height:calc-(100)",
      "type": "utility",
      "important": false
    }
  ]
Tools / parse bad class with bracket / bad bracket / 0: |-
  [
    {
      "raw": "xyz()",
      "start": 0,
      "end": 5,
      "variants": [],
      "content": "xyz()",
      "type": "utility",
      "important": false
    },
    {
      "raw": "p-5",
      "start": 6,
      "end": 9,
      "variants": [],
      "content": "p-5",
      "type": "utility",
      "important": false
    },
    {
      "raw": "bg-red-500",
      "start": 10,
      "end": 20,
      "variants": [],
      "content": "bg-red-500",
      "type": "utility",
      "important": false
    },
    {
      "raw": "xyz-()",
      "start": 21,
      "end": 27,
      "variants": [],
      "content": "xyz-()",
      "type": "utility",
      "important": false
    },
    {
      "raw": "xy-(",
      "start": 28,
      "end": 32,
      "variants": [],
      "content": "xy-(",
      "type": "utility",
      "important": false
    },
    {
      "raw": "xy)",
      "start": 33,
      "end": 36,
      "variants": [],
      "content": "xy)",
      "type": "utility",
      "important": false
    },
    {
      "raw": "bg-yellow-600",
      "start": 37,
      "end": 50,
      "variants": [],
      "content": "bg-yellow-600",
      "type": "utility",
      "important": false
    }
  ]
Tools / parse bad class with half bracket / bad half bracket / 0: |-
  [
    {
      "raw": "abc(",
      "start": 0,
      "end": 4,
      "variants": [],
      "content": "abc(",
      "type": "utility",
      "important": false
    },
    {
      "raw": "text-lg",
      "start": 5,
      "end": 12,
      "variants": [],
      "content": "text-lg",
      "type": "utility",
      "important": false
    },
    {
      "raw": "font-bold",
      "start": 13,
      "end": 22,
      "variants": [],
      "content": "font-bold",
      "type": "utility",
      "important": false
    }
  ]
Tools / parse bad class with quotes / bad classes / 0: |-
  [
    {
      "raw": "${",
      "start": 0,
      "end": 2,
      "variants": [],
      "content": "${",
      "type": "utility",
      "important": false
    },
    {
      "raw": "active",
      "start": 3,
      "end": 9,
      "variants": [],
      "content": "active",
      "type": "utility",
      "important": false
    },
    {
      "raw": "?",
      "start": 10,
      "end": 11,
      "variants": [],
      "content": "?",
      "type": "utility",
      "important": false
    },
    {
      "raw": "text-green-400",
      "start": 13,
      "end": 27,
      "variants": [],
      "content": "text-green-400",
      "type": "utility",
      "important": false
    },
    {
      "raw": ":",
      "start": 29,
      "end": 30,
      "variants": [],
      "content": ":",
      "type": "utility",
      "important": false
    },
    {
      "raw": "text-orange-400",
      "start": 32,
      "end": 47,
      "variants": [],
      "content": "text-orange-400",
      "type": "utility",
      "important": false
    },
    {
      "raw": "}",
      "start": 49,
      "end": 50,
      "variants": [],
      "content": "}",
      "type": "utility",
      "important": false
    }
  ]
Tools / parse function with parameter deconstruction / jsx function component / 0: |-
  [
    {
      "raw": "function",
      "start": 0,
      "end": 8,
      "variants": [],
      "content": "function",
      "type": "utility",
      "important": false
    },
    {
      "raw": "ComponentA({",
      "start": 9,
      "end": 21,
      "variants": [],
      "content": "ComponentA({",
      "type": "utility",
      "important": false
    },
    {
      "raw": "name",
      "start": 22,
      "end": 26,
      "variants": [],
      "content": "name",
      "type": "utility",
      "important": false
    },
    {
      "raw": "}",
      "start": 27,
      "end": 28,
      "variants": [],
      "content": "}",
      "type": "utility",
      "important": false
    },
    {
      "raw": "Props",
      "start": 29,
      "end": 34,
      "variants": [],
      "content": "Props",
      "type": "utility",
      "important": false
    },
    {
      "raw": "{",
      "start": 36,
      "end": 37,
      "variants": [],
      "content": "{",
      "type": "utility",
      "important": false
    },
    {
      "raw": "return",
      "start": 38,
      "end": 44,
      "variants": [],
      "content": "return",
      "type": "utility",
      "important": false
    },
    {
      "raw": "div",
      "start": 45,
      "end": 48,
      "variants": [],
      "content": "div",
      "type": "utility",
      "important": false
    },
    {
      "raw": "className=",
      "start": 49,
      "end": 59,
      "variants": [],
      "content": "className=",
      "type": "utility",
      "important": false
    },
    {
      "raw": "grid",
      "start": 60,
      "end": 64,
      "variants": [],
      "content": "grid",
      "type": "utility",
      "important": false
    }
  ]
Tools / parse important / parse important / 0: |-
  [
    {
      "raw": "!text-green-300",
      "start": 0,
      "end": 15,
      "variants": [],
      "content": "text-green-300",
      "type": "utility",
      "important": true
    },
    {
      "raw": "font-bold",
      "start": 16,
      "end": 25,
      "variants": [],
      "content": "font-bold",
      "type": "utility",
      "important": false
    },
    {
      "raw": "!hover:(p-4 bg-red-500)",
      "start": 26,
      "end": 49,
      "variants": [
        "hover"
      ],
      "content": [
        {
          "raw": "p-4",
          "start": 34,
          "end": 37,
          "variants": [],
          "content": "p-4",
          "type": "utility",
          "important": false
        },
        {
          "raw": "bg-red-500",
          "start": 38,
          "end": 48,
          "variants": [],
          "content": "bg-red-500",
          "type": "utility",
          "important": false
        }
      ],
      "type": "group",
      "important": true
    },
    {
      "raw": "focus:(!border float-right)",
      "start": 50,
      "end": 77,
      "variants": [
        "focus"
      ],
      "content": [
        {
          "raw": "!border",
          "start": 57,
          "end": 64,
          "variants": [],
          "content": "border",
          "type": "utility",
          "important": true
        },
        {
          "raw": "float-right",
          "start": 65,
          "end": 76,
          "variants": [],
          "content": "float-right",
          "type": "utility",
          "important": false
        }
      ],
      "type": "group",
      "important": false
    }
  ]
